lib/galaxy/datatypes/converters/fasta_to_2bit.xml

```diff
@@ -2,8 +2,7 @@
     <!-- <description>__NOT_USED_CURRENTLY_FOR_CONVERTERS__</description> -->
     <!-- Used on the metadata edit page. -->
     <requirements>
-        <requirement type="package" version="332">ucsc-fatotwobit</requirement>
```

Comment:
If we don't want to be too tied to conda packages, I can live with the `ucsc_tools` requirement name (for which IUC has already created a conda metapackage), even if this name is not used anywhere in http://hgdownload.soe.ucsc.edu/admin/exe/ . But, as I've said on Gitter, I think we should add (not remove) version numbers in the tool `<requirement>`s, to have better reproducibility.

Reply:
I think it's at least good to track which individual UCSC binaries by each tool since it's fairly common to not install them all. Maybe a requirement tag is not the right way to do this though.

Reply:
Wait... this comment wasn't really addressed IMO. @nsoranzo mentioned the version and @natefoo responded

> I think it's at least good to track which individual UCSC binaries by each tool 

This tool had the individual binary referenced already 😓.

Reply:
Sorry I did not see this before merging @nsoranzo. Do you want to open a PR?

Reply:
I disagree here with @nsoranzo, I prefer this variant so I +1'd this but I knew there was a pretty fundamental disagreement so I didn't merge. I'm not sure what to do longer term with people having such differing opinions on what requirements should represent and express.

Reply:
@jmchilton Can you elaborate how not having version numbers is better? Having them is quite clearly better for reproducibility.

Reply:
@nsoranzo I feel like we keep having the same conversation we just don't agree about these things at a very fundamental level 😢, is more conversation going to help?

> Can you elaborate how not having version numbers is better?

It says ``requirement``, these tools literally do not have a requirement on any particular version of these applications.

> Having them is quite clearly better for reproducibility.

This claim is an opinion and comes down to rather subjective views on what reproducibility means. If two researchers download a new Galaxy, configure Conda dependencies, and install these tools they will both get the same version of primary dependency (though perhaps differing versions of the dependencies' dependencies depending on the time they install). If instead we added ``version`` to the requirements the same holds true. So the version bought us nothing really here because of the Conda mapping files.

If instead, I have a Galaxy instance that I've had for a long time and I've configured my system to map these requirements to environment modules and my users have used these tools in the past and now the Galaxy team adds versions to the requirements - I'll probably want to upgrade the environment modules rather than point them at conflicting versions and then results for seemingly the same tool run will change slightly. We've decreased the re-computability over time of these tools by introducing an artificial version - *we've decreased the reproducibility*, we've decreased the remix-ability (just as important).

Can one imagine a scenario and set of metrics for defining "reproducibility" for which adding the versions helps? Of course! I'm not claiming my definition or intuition is obviously better, just that it is subjective. In light of it being subjective, I think my strong desire to not rely directly on Conda in tool definitions when we can avoid it (tools should be abstract and should support many dependency resolvers) and my desire to avoid artificial, inaccurate assertions about the versions of applications required by tools lends me to believe that annotating versions on these requirements is the wrong thing to do.

Reply:
> So the version bought us nothing really here because of the Conda mapping files.

That's not true, not having the version attribute for a tool requirement means that:
- if you don't have an entry for this requirement in `default_conda_mapping.yml`, then if a new version changes the semantic of some option you're screwed (think about the `-o` option of `samtools index` which changed between 0.1.19 and 1.0)
- if instead there's an entry for it in `default_conda_mapping.yml`, then you cannot have 2 tools which don't specify the requirement version, but actually require different ones.

In short, you cannot assume that a tool requirement won't change its semantic in the future, so you need to specify the version where the tool is known to work.

> We've decreased the re-computability over time of these tools by introducing an artificial version - we've decreased the reproducibility

No, we have increased the reproducibility for future tool executions, what hinders the reproducibility of previous runs is the lack of version.

> I think my strong desire to not rely directly on Conda in tool definitions when we can avoid it (tools should be abstract and should support many dependency resolvers)

The need for requirement versions is not specific to conda (or environment modules).

Reply:
@martenson https://github.com/galaxyproject/galaxy/pull/5973

Reply:
> then you cannot have 2 tools which don't specify the requirement version

But this is not the case for these requirements and will never be the case. My imagined scenario seems more likely than yours for this particular set of requirements.

> you cannot assume that a tool requirement won't change its semantic in the future, so you need to specify the version where the tool is known to work.

The mapping file approach doesn't derive its reproducibility from not assuming versions won't change in the future - it uses pins for a reason. You are making an argument about an expression gap here I think not an argument about reproducibility with this line. I'll readily admit there are totally issues with the mapping files - but preemptive stuff like #5973 isn't the right way out of it IMO. As we encounter cases where the expression gap hits us hard, we should be building infrastructure for more expressive tools that are equally reproducible in your sense (e.g. per-tool Conda mapping files paired with version range support for requirements), not making tools less expressive by limiting the versions of requirements they can target.

Like I said we are not going to reach an agreement here right? I don't really get the sense you care at all about the same things I do in terms of reproducibility and tool expression. You have your own view point and set of use cases you care about and I totally respect that and understand your reasoning. The result just kind of makes me feel hopeless.

Reply:
I am with @nsoranzo on this one, I think annotating the requirements with versions known to work is the right thing to do since commandline options as well as results can and have changed previously (samtools is a great example! It was broken for new instances for more than a year because no-one noticed). If admins didn't disable the versionless resolvers nothing will change anyway, or am I missing something ? To me this is future-facing reproducibility, and we should aim for that with the option to opt out (=versionless), not the other way round where you can opt-in by mapping versionless to a specific version.